controllers/machine_controller_test.go

```diff
@@ -32,11 +32,146 @@ import (
 	"sigs.k8s.io/cluster-api/test/helpers"
 	"sigs.k8s.io/cluster-api/util"
 	"sigs.k8s.io/cluster-api/util/conditions"
+	"sigs.k8s.io/cluster-api/util/patch"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
+func TestWatches(t *testing.T) {
+	g := NewWithT(t)
+	ns, err := testEnv.CreateNamespace(ctx, "test-machine-watches")
+	g.Expect(err).ToNot(HaveOccurred())
+	defer func() {
+		g.Expect(testEnv.Cleanup(ctx, ns)).To(Succeed())
```

Comment:
This unfortunately does not actually fully remove the namespace and all its dependents from etcd. testEnv doesn't run kube-controller-manager, which is required for full removal of a namespace and all its dependents. You'll have to call Cleanup for every object you create.